### PR TITLE
[BUG] On slave, expiref docs do not return NULL

### PR DESCRIPTION
**Pull request by ashtul**
On slave shard, docs does not `expire` as redisearch attempt to load it.
This means `Document_Deleted` flag is not set and the reply array is either empty or contains SORTABLE fields.

If loading has failed, we attempt to remove the document from the index.

The test is stable using `DEBUG SET-ACTIVE-EXPIRE 0`. Documents now expire only when we try to access them, so the first time we call `FT.SEARCH`, we should find the document in the DocTable but will fail to load the fields and return (nil).

Related to #2651 
MOD-3515

### CODE REVIEW

**Review by MeirShpilraien — COMMENTED** [review]
Look good, I believe we can extend the test to also have a replication and do the search on the replica as well (before doing the search on the primary) so we will see that the replica also act as expected.

**Comment by ashtul** [comment]
@MeirShpilraien replication test added + handling redis 6 vs 7

File: tests/pytests/test_expire.py [hunk 1]
```diff
@@ -36,3 +38,26 @@ def testExpire(env):
             time.sleep(1)
     except Exception as e:
         env.assertEqual(str(e), 'Unknown index name')
+
+def testExpireDocs(env):
+    env.skipOnCluster()
+    if server_version_at_least(env, '7.0.0'):
```
Comment (MeirShpilraien) [inline tests/pytests/test_expire.py:44]: Why is it needed?

File: tests/pytests/test_replicate.py [hunk 2]
```diff
@@ -41,6 +41,8 @@ def checkSlaveSynced(env, slaveConn, command, expected_result, time_out=5):
 
 def initEnv():
   env = Env(useSlaves=True, forceTcp=True)
+  if server_version_at_least(env, '7.0.0'):
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:44]: Why?

File: tests/pytests/test_replicate.py [hunk 3]
```diff
@@ -41,6 +41,8 @@ def checkSlaveSynced(env, slaveConn, command, expected_result, time_out=5):
...
+    env = Env(useSlaves=True, forceTcp=True, enableDebugCommand=True)
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:45]: You are potentially creating 2 envs at the same test!!

File: tests/pytests/test_replicate.py [hunk 4]
```diff
@@ -211,3 +213,38 @@ def testDropWith__FORCEKEEPDOCS():
...
+def testExpireDocs():
+  env = initEnv()
+  master = env.getConnection()
+  slave = env.getSlaveConnection()
+
+  '''
+  This test creates creates an index and two documents and check they
+  exist on both shards.
+  One of the documents is expired.
+  The test checks the document is removed from both master and slave.
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:226]: I do not see you use: `'DEBUG', 'SET-ACTIVE-EXPIRE', '0'`
The idea is to check active expire on slave as well

**Review by MeirShpilraien — COMMENTED** [review]
Added some comments

File: tests/pytests/test_replicate.py [hunk 5]
```diff
@@ -211,3 +213,38 @@ def testDropWith__FORCEKEEPDOCS():
...
+  
+  # both docs exist
+  res = master.execute_command('FT.SEARCH', 'idx', '*')
+  env.assertEqual(res, [2, 'doc1', ['t', 'foo'], 'doc2', ['t', 'bar']])
+  checkSlaveSynced(env, slave, ('FT.SEARCH', 'idx', '*', 'NOCONTENT'), [2, 'doc1', 'doc2'], time_out=5)
+
+  master.execute_command('PEXPIRE', 'doc1', 1)
+  time.sleep(0.01)
+
+  # both docs exist but doc1 fail to load field since they were expired passively
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:241]: Search first on slave to check passive expiration on slave works as expected (no crashes).

File: tests/pytests/test_replicate.py [hunk 6]
```diff
@@ -211,3 +211,85 @@ def testDropWith__FORCEKEEPDOCS():
...
+  for i in range(2):
+    sortby_cmd = [] if i == 0 else ['SORTBY', 't']
+    master.execute_command('FT.CREATE', 'idx', 'SCHEMA', 't', 'TEXT')
+    master.execute_command('HSET', 'doc1', 't', 'bar')
+    master.execute_command('HSET', 'doc2', 't', 'foo')
+    
+    # both docs exist
+    res = master.execute_command('FT.SEARCH', 'idx', '*', *sortby_cmd)
+    env.assertEqual(res, [2, 'doc1', ['t', 'bar'], 'doc2', ['t', 'foo']])
+    checkSlaveSynced(env, slave, ('FT.SEARCH', 'idx', '*', 'NOCONTENT'), [2, 'doc1', 'doc2'], time_out=5)
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:239]: Any reason why you run with `NOCONTENT` here?

File: tests/pytests/test_replicate.py [hunk 7]
```diff
@@ -211,3 +211,85 @@ def testDropWith__FORCEKEEPDOCS():
...
+
+    master.execute_command('PEXPIRE', 'doc1', 1)
+    time.sleep(0.01)
+
+    # since there is no sortable, we loaded doc1 at sortby and found out it was deleted
+    res = slave.execute_command('FT.SEARCH', 'idx', '*', *sortby_cmd)
+    env.assertEqual(res, [1, 'doc2', ['t', 'foo']])
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:246]: But what if we do not use SORTBY on the query? How come we did not get NULL result?

File: tests/pytests/test_replicate.py [hunk 8]
```diff
@@ -211,3 +211,85 @@ def testDropWith__FORCEKEEPDOCS():
...
+    res = master.execute_command('FT.SEARCH', 'idx', '*', *sortby_cmd)
+    env.assertEqual(res, [1, 'doc2', ['t', 'foo']])
+
+    master.execute_command('FLUSHALL')
+    checkSlaveSynced(env, slave, ('KEYS','*'), [], time_out=5)
```
Comment (MeirShpilraien) [inline tests/pytests/test_replicate.py:251]: Why not using `wait` here?

File: src/rlookup.c [hunk 9]
```diff
@@ -538,8 +538,8 @@ static int RLookup_HGETALL(RLookup *it, RLookupRow *dst, RLookupLoadOptions *opt
   // and when the deployment is not enterprise-crdt
   if(!isFeatureSupported(RM_SCAN_KEY_API_FIX) || isCrdt){
     rep = RedisModule_Call(ctx, "HGETALL", "s", krstr);
-
     if (rep == NULL || RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_ARRAY) {
+      QueryError_SetCode(options->status, QUERY_ENODOC);
```
Comment (MeirShpilraien) [inline src/rlookup.c:542]: Consider removing this and count on the return value

**Review by MeirShpilraien — COMMENTED** [review]
Looks good :+1: 
Some small comments/questions

File: tests/pytests/test_replicate.py (hunk 6 above)
Comment (ashtul) [reply tests/pytests/test_replicate.py:239]: We don't expect a change and IMO it is more readable.
Changed

File: src/rlookup.c (hunk 9 above)
Comment (ashtul) [reply src/rlookup.c:542]: Using now the return value
This is inline with other load functions such as `getKeyCommonHash`

File: tests/pytests/test_replicate.py (hunk 7 above)
Comment (ashtul) [reply tests/pytests/test_replicate.py:246]: fixed

File: src/result_processor.c [hunk 10]
```diff
@@ -452,7 +452,9 @@ static int rpsortNext_innerLoop(ResultProcessor *rp, SearchResult *r) {
                                     .status = &status};
       RLookup_LoadDocument(NULL, &h->rowdata, &loadopts);
       if (QueryError_HasError(&status)) {
-        return RS_RESULT_ERROR;
+        // failure to fetch the doc. ignore it and continue 
+        rp->parent->totalResults--;
+        return RESULT_QUEUED;
```
Comment (MeirShpilraien) [inline src/result_processor.c:457]: Thinking about it now, what if someone will do sortby on a none existing field? Before he will get an error and now we will skip the document?

File: src/rlookup.c (hunk 9 above)
Comment (MeirShpilraien) [reply src/rlookup.c:542]: So why you still need to set the error code here? I do not mind rearrange stuff but lets do it on separate PR, remember that this PR need to be cherrypicked.

File: src/result_processor.c (hunk 10 above)
Comment (ashtul) [reply src/result_processor.c:457]: That is true.
Lets let it get into the queue. and we should add a test for it

Comment (MeirShpilraien) [reply src/result_processor.c:457]: Summery of the discussion:
If the field does not exists the `loadIndividualKeys` would have return REDISMODULE_OK and will not return with error. So it is not a breaking change.

**Review by MeirShpilraien — DISMISSED** [review]

**Review by MeirShpilraien — DISMISSED** [review]

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**